table/block_based_table_reader.cc

```diff
@@ -1499,7 +1501,30 @@ InternalIterator* BlockBasedTable::NewDataBlockIterator(
     assert(block.value != nullptr);
     iter = block.value->NewIterator(&rep->internal_comparator, input_iter, true,
                                     rep->ioptions.statistics);
-    if (block.cache_handle != nullptr) {
+    if (!ro.fill_cache && rep->cache_key_prefix_size != 0) {
+      // insert a dummy record to block cache to track the memory usage
+      Cache::Handle* cache_handle;
+      // There are two other types of cache keys: 1) SST cache key added in
+      // `MaybeLoadDataBlockToCache` 2) dummy cache key added in
+      // `write_buffer_manager`. Use longer prefix (41 bytes) to differentiate
+      // from SST cache key(31 bytes), and use non-zero prefix to differentiate
+      // from `write_buffer_manager`
+      const size_t kExtraCacheKeyPrefix = kMaxVarint64Length * 4 + 1;
+      char cache_key[kExtraCacheKeyPrefix + kMaxVarint64Length];
+      // Prefix: use rep->cache_key_prefix padded by 0s
+      memset(cache_key, 0, kExtraCacheKeyPrefix + kMaxVarint64Length);
+      assert(rep->cache_key_prefix_size != 0);
+      assert(rep->cache_key_prefix_size <= kExtraCacheKeyPrefix);
+      memcpy(cache_key, rep->cache_key_prefix, rep->cache_key_prefix_size);
+      char* end =
+          EncodeVarint64(cache_key + kExtraCacheKeyPrefix, next_cache_key_id_++);
+      assert(end - cache_key <= int (kExtraCacheKeyPrefix + kMaxVarint64Length));
+      Slice unique_key = Slice(cache_key, static_cast<size_t>(end - cache_key));
+      block_cache->Insert(unique_key, nullptr, block.value->size(),
+                          nullptr, &cache_handle);
+      iter->RegisterCleanup(&ReleaseCachedEntry, block_cache, cache_handle);
```

Comment:
Do we need to do the work done in DeleteHeldResource() here?

Reply:
I think calling cache->Release(handle, true) is sufficient here? 
If we delete the block here, the block might be double deleted later